src/components/implementation/tests/micro_booter/mb_tests.c

```diff
@@ -487,6 +491,197 @@ test_budgets(void)
+	 * expected result:
+	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
+	 * - cos_rcv from intr_thd should activate worker_thd.
+	 */
+	PRINTC(" Test - Intr = H, Worker = M, Preempted = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_LOW, TEST_PRIO_HIGH, TEST_PRIO_MED);
+
+	/*
+	 * test case 2: worker_thd = H, intr_thd = M, preempted_thd = L
+	 * expected result:
+	 * - cos_thd_wakeup should invalidate preempted_thd and add worker_thd there.
+	 * - cos_rcv from intr_thd should activate worker_thd.
+	 */
+	PRINTC(" Test - Worker = H, Intr = M, Preempted = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_LOW, TEST_PRIO_MED, TEST_PRIO_HIGH);
+
+	/*
+	 * test case 3: intr_thd = H, preempted_thd = M, worker_thd = L
+	 * expected result:
+	 * - cos_thd_wakeup should not invalidate preempted_thd.
+	 * - cos_rcv from intr_thd should activate preempted_thd.
+	 */
+	PRINTC(" Test - Intr = H, Preempted = M, Worker = L\n");
+	test_wakeup_case(&wat, TEST_PRIO_MED, TEST_PRIO_HIGH, TEST_PRIO_LOW);
+
+	wakeup_test_start = 0;
+	PRINTC("Done.\n");
+}
+
+static void
+receiver_thd(void *d)
+{
+	struct exec_cluster *e = &(((struct activation_test_data *)d)->w);
+
+	while (1) {
+		cos_rcv(e->rc);
+		PRINTC("\tReceiver: activated\n");
+	}
+}
+
+static void
+sender_thd(void *d)
+{
+	struct exec_cluster *e = &(((struct activation_test_data *)d)->i);
+	struct exec_cluster *r = &(((struct activation_test_data *)d)->w);
+
+	while (1) {
+		cos_asnd(r->sc, 0);
+		PRINTC("\tSender: activated\n");
+		cos_rcv(e->rc);
+	}
+}
+
+static void
+test_preemption_case(struct activation_test_data *at, tcap_prio_t iprio, tcap_prio_t wprio)
+{
+	at->i.prio = iprio;
+	at->w.prio = wprio;
+	if (cos_tcap_transfer(at->i.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, at->i.prio)) assert(0);
+	if (cos_tcap_transfer(at->w.rc, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, at->w.prio)) assert(0);
+	cos_switch(at->i.tc, at->i.tcc, at->i.prio, TCAP_TIME_NIL, 0, 0);
+}
+
+static void
+test_preemption(void)
+{
+	PRINTC("Testing Preemption\n");
+	exec_cluster_alloc(&pat.i, sender_thd, &pat, BOOT_CAPTBL_SELF_INITRCV_BASE);
+	exec_cluster_alloc(&pat.w, receiver_thd, &pat, BOOT_CAPTBL_SELF_INITRCV_BASE);
+
+	/*
+	 * test case 1: Sender = H, Receiver: L 
+	 * - scheduler here is initthd (this thd)
+	 * expected result:
+	 * - cos_asnd from sender should add receiver as wakeup thread.
+	 * - cos_rcv from sender should activate receiver thread.
+	 */
+	PRINTC(" Test - Sender = H, Receiver = L\n");
+	test_preemption_case(&pat, TEST_PRIO_HIGH, TEST_PRIO_LOW);
+
+	/*
+	 * test case 2: Sender = L, Receiver: H 
+	 * - scheduler here is initthd (this thd)
+	 * expected result:
+	 * - cos_asnd from sender should trigger receiver activation and add sender as wakeup thread.
+	 * - cos_rcv from receiver should activate sender thread.
+	 */
+	PRINTC(" Test - Sender = L, Receiver = H\n");
+	test_preemption_case(&pat, TEST_PRIO_LOW, TEST_PRIO_HIGH);
```

Comment:
You did a great job abstracting away the shared parts of the code, but I'm missing the code that checks that the proper control flow for each case actually happens.